---
title: "Transcription Factor Hierarchy"
author: "Jose Americo N L F Freitas"
date: "July 3, 2018"
output:
   BiocStyle::html_document:
    theme: cosmo
    highlight:  "tango"
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 2
    number_sections: true
    code_folding: 'show'
    df_print: paged
abstract: |
  This Rmarkdown file contains the software that generates the panels in Figure 5. We start generating transcription factor hierarchy networks for each one of the gene modules obatined by the WGCNA analysis on a transcriptome timecourse of RAS-induced senescence on WI-38 fibroblasts.
vignette: |
  %\VignetteIndexEntry{Authoring R Markdown vignettes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: bibliography.bib 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE, eval = TRUE, results = "hide", message = FALSE, warning = FALSE, echo = TRUE, fig.wide = TRUE,  fig.align='center')
```

# Introduction

In order to understand how transcription factors act collectivelly to regulate transcription, we generated networks depicting they spatial precedence [@garber2012] during oncogene-induced senescence (OIS). This document contains the workflow used to generate those networks and calls several scripts at the *function* folder.


# Building the Transcription Factor Hierachy Networks

## Importing data

The input data in this instance consists of a list of seven TF *social maps*. Each *social map* corresponds to a transcriptional cluster obtained by the WGCNA analysis and consists of a matrix contaning TF positional information, with rows representing gene enhancers and columns, TFs. Non-zero matrix elements mean that PIQ has inferred at least one peak for TF at the region corresponding to the current row.

The *social maps* were generated by the script described in *script of Figure 3*. To save processing time, they are loaded from the *data* folder. All TFs with similar PWMs (overlap > 80%) are grouped together to reduce the false positive rate,.

```{r}

  # Data and source directories
  srcFolder = paste0(getwd(), '/function')
  dataFolder = paste0(getwd(), '/data')
  
  # Social_Map_Final: a variable containing all TFs positional information
  fileSocialMaps = '/RAS_OIS_GEM_SOCIAL_MAPS_ENHANCERS_W250_NO5.RData'
  load(paste0(dataFolder, fileSocialMaps))
  
  source(paste0(srcFolder, '/5A_01_TFH_PREPROCESSING.R'))
  tfClustersList = tfCluster(dataFolder, Social_Map_Final)
  SocialMapFinalPWM = socialMapCluster(Social_Map_Final, tfClustersList)

```

## Computing precedence relationships and TF activity data

All precedence relationships are considered at this stage. The *tfHierarchy()* function goes through each cis-regulatory region and adds the region index to a list corresponding to all cobinding instances of a given TF pair. This is performed by updating an adjacency matrix which elements are lists containing the regions indexes where TF i appeared at the same instant or before TF j. Hence, the number of regions that two TFs interact over time is given by the length of the list at element (i, j).

This function also keeps track of the first day each interaction happened, which will be used for normalization purposes in the next sections.

```{r}

  source(paste0(srcFolder, '/5A_02_TFH_HIERARCHY.R'))
  tfNetList = tfHierarchy(SocialMapFinalPWM)
  
```

## Computing activity parameters & normalizing networks

### Activity parameters

In order to study TF properties, several metrics were computed:

* Dynamicity
* Total number of bound regions
* Prestimulation (number of bound regions at D0)
* Total number of regions (per day)

All formulas are described in Supplementary methods.

```{r}

  source(paste0(srcFolder, '/5A_03_TFH_ACTIVITY.R'))
  tfActivity = tfActivityFactors(SocialMapFinalPWM)
  tfInfoList = activityAsDataFrame(tfActivity)
  
```

### Normalization

In order to account for the difference in the number of reads for each timepoint, the intensity of each edge is calculated as a weighted sum of regions. Each term is normalized by the first day the cobinding happened at that region.

```{r}
  
  source(paste0(srcFolder, '/5A_04_TFH_NORMALIZATION.R'))
  tfNetNormalizedList = tfNetNormalize(tfNetList, tfActivity)
  
```

## Network operations

In order to improve network visualization and processing, we performed four operations on each one:

* Filtering: we keep only edges corresponding to precedence relationships that cover more than 30 % of the target TF regions. 
* Strong connected components (SCC): in graph theory, a strong connected component (SCC) is defined as a set of nodes that can be accessed from each other. In the context of TFHs, a SCC comprises TFs that precede each other during the time course. The implementation on this project was described by @tarjan1972 .
* Vertex Sort [VS; @jothi2009]: this algorithm performs a topological sort by finding the distance of each node to the network top and bottom. Since it relies on the network SCCs to avoid cycles, it receives both the networks and their SCCs as parameters.
* Transitive reduction [TR; @aho1972]: when a TF A precedes TF B and TF B precedes TF C, it is likely that TF A precedes TF C (if the overlap product from the previous interactions plus the direct precedence of TF A to TF C exceeds 30 %). With the aim of simplifying the network, the indirect edges A -> C are prunned. Our approach involves the computation of the networks Transitive Closure (TC), which links every node to all nodes accessible from it [@aho1972], and subsequent checking for nodes that are acessible by the current node's neighbors (caracterizing the indirect relationships). In order to optmize the script's execution, the TC is performed at the same time as the SCC algorithm, as described by @nuutila1993.

```{r}
  FILTER_THRESHOLD = .3

  source(paste0(srcFolder, '/5A_05_TFH_FILTER.R'))
  tfNetFilteredList = filterThreshold(tfNetNormalizedList, FILTER_THRESHOLD)
  
  source(paste0(srcFolder, '/5A_06_TFH_SCC.R'))
  tfNetFilteredSccList = netSCC(tfNetFilteredList)
  
  source(paste0(srcFolder, '/5A_07_TFH_VS.R'))
  tfNetFilteredVsList = netVS(tfNetFilteredList, tfNetFilteredSccList)
  
  source(paste0(srcFolder, '/5A_08_TFH_TR.R'))
  tfNetFilteredTrList = netTR(tfNetFilteredList, tfNetFilteredSccList)
  tfNetFilteredSccTrList = netTR(tfNetFilteredSccList, tfNetFilteredSccList)

```

# Network visualization

We save all networks as *igraph* objects, where each TF activity parameter is registered as node attributes and edges are weighted. These objects will be used for the analises described in all other panels.

```{r, message = FALSE}

  source(paste0(srcFolder, '/5A_09_TFH_DATA.R'))
  tfGraphFilteredList = graphData(dataFolder, tfNetFilteredList, tfNetFilteredVsList$vertexLayersList, tfInfoList)
  tfGraphFilteredSccList = graphData(dataFolder, tfNetFilteredSccList, tfNetFilteredVsList$vertexLayersSccList, tfInfoList)
  tfGraphFilteredTrList = graphData(dataFolder, tfNetFilteredTrList, tfNetFilteredVsList$vertexLayersList, tfInfoList)
  tfGraphFilteredSccTrList = graphData(dataFolder, tfNetFilteredSccTrList, tfNetFilteredVsList$vertexLayersSccList, tfInfoList)
```

We visualize the networks with CytoScape [@cytoscape] software with the use of CyREST [@cyrest] interface and two packages: R2Cytoscape [@r2cytoscape] and RCy3 [@rcy3].

The following code is not automatically executed when running this Rmarkdown file. You should first open CytoScape and then run it. While the script is running, you'll see the networks being created on CytoScape's window.

```{r, message = FALSE, eval = FALSE}

  # CytoScape session control
  source(paste0(srcFolder, '/5A_10_TFH_CYTOSCAPE.R'))
  cytoscapeSession = list(portNumber = 1234,
                        baseUrl = paste0("http://localhost:", toString(1234), "/v1"))

  # Generating network files
  lapply(seq(length(tfGraphFilteredList)), function(c){
    newSession(cytoscapeSession$baseUrl, destroyCurrent = TRUE)

    graphREST(cytoscapeSession, tfGraphFilteredList[[c]], netName = paste('tfh', 'enhancer', c, sep = '_'), srcFolder = srcFolder)
    graphREST(cytoscapeSession, tfGraphFilteredSccList[[c]], paste('tfh', 'enhancer', 'scc', c, sep = '_'), srcFolder = srcFolder)
    graphREST(cytoscapeSession, tfGraphFilteredTrList[[c]], paste('tfh', 'enhancer', 'tr', c, sep = '_'), srcFolder = srcFolder)
    graphREST(cytoscapeSession, tfGraphFilteredSccTrList[[c]], paste('tfh', 'enhancer', 'scc', 'tr', c, sep = '_'), srcFolder = srcFolder)

    saveSessionToFile(cytoscapeSession$baseUrl, paste0(dataFolder, '/tfh_enhancer_', c))
  })

```

# Incoming edges classification distribution

We analyze TF precedence relationship focusing on their classification, e.g., the proportion of settlers succeding pioneers or migrants. The results of this analysis are displayed in Figure 5D.

Considering that TFs have three possible classifications ('pioneer', 'settler', 'migrant') and that edges link a 'source' to a 'target', we have nine different classes of edges.

```{r}
  library(igraph)

  # Extracting adjacency matrices and classification information
  tfGraphFilteredAdjMatList = lapply(tfGraphFilteredList, function(tfG) igraph::as_adj(tfG))
  tfClassif = V(tfGraphFilteredList[[1]])$classif

  # 
  pioneerIndexes = (tfClassif == 'p')
  settlerIndexes = (tfClassif == 's')
  migrantIndexes = (tfClassif == 'm')

  listIndexes = list(pioneer = pioneerIndexes, settler = settlerIndexes, migrant = migrantIndexes)
  
  # Counts the number of edges with same target and source classification
  edgeClasses = lapply(tfGraphFilteredAdjMatList, function(tfN)
    sapply(listIndexes, function(parentNode) 
      sapply(listIndexes, function(childNode) 
        sum(tfN[parentNode, childNode])
      )
    )
  )
  
  # Considering all modules
  edgeClassesAllModules = Reduce('+', edgeClasses)
  
  # Properly naming rows and columns
  rownames(edgeClassesAllModules) = paste(rownames(edgeClassesAllModules), 'as target')
  colnames(edgeClassesAllModules) = paste(colnames(edgeClassesAllModules), 'as source')
  
```

As depicted in panel 5D, we calculate both absolute and relative number of edges for each TF classification.

```{r, results = "markup"}
  # Printing absolute number of edges
  edgeClassesAllModules
  
  # Printing relative number of edges
  round(sweep(edgeClassesAllModules, 1, rowSums(edgeClassesAllModules), '/'), 2)
```

# TF properties

Pannel 5E shows dynamicity and number of bound regions for each TF with respect to each gene transcriptional module.

## Preprocessing data

We scale the TF parameters (dynamicity and number of bound regions) in order to highlight distinct TF properties. For better visualization, we normalize the 'number of bound regions' parameter with the function *sqrt* before scaling.

```{r, message = FALSE}
  library(igraph)
  library(ComplexHeatmap)
  library(circlize)
  library(vegan)

  tfLayerPropertiesSet = lapply(tfGraphFilteredList, function (tfG) 
    as.data.frame(vertex_attr(tfG)[c('realName', 'layer', 'dynamic_norm', 'classif', 'nWin_total_norm', 'prestim_norm')]))

  layerNames = c('top', 'complexMiddle', 'simpleMiddle', 'bottom')
  classifNames = c('p', 's', 'm')

  #
  tfLayerDynamicsScaled = Reduce('cbind', lapply(seq(length(tfLayerPropertiesSet)), function(c){
    x = tfLayerPropertiesSet[[c]]
    df = data.frame(matrix(nrow = nrow(tfLayerPropertiesSet[[c]]), ncol = 0))
    df$realName = factor(tfLayerPropertiesSet[[c]]$realName, levels = tfLayerPropertiesSet[[c]]$realName)
    df$dynamic_scaled = scale(tfLayerPropertiesSet[[c]]$dynamic_norm)
  }))

  #
  tfLayerNRegScaled = Reduce('cbind', lapply(seq(length(tfLayerPropertiesSet)), function(c){
    x = tfLayerPropertiesSet[[c]]
    df = data.frame(matrix(nrow = nrow(tfLayerPropertiesSet[[c]]), ncol = 0))
    df$realName = factor(tfLayerPropertiesSet[[c]]$realName, levels = tfLayerPropertiesSet[[c]]$realName)
    df$nReg_scaled = scale(sqrt(tfLayerPropertiesSet[[c]]$nWin_total_norm))
  }))

  #
  tfLayerClassif = tfLayerPropertiesSet[[1]]$classif

```

## Hierarchical clustering

We cluster the TFs based on their properties for each transcriptional module. In order to assess the relationship between those parameters and TF classification, we sort the dendrogram accordingly to the ratio between dynamicity and coverage.

```{r}
# Merge all data for hierarchical clustering
allDataHeatMap = cbind(tfLayerDynamicsScaled, tfLayerNRegScaled)

# Clustering
hCluster = hclust(dist(allDataHeatMap, method = 'euclidean'), method = 'complete')

# Reordering
reorderWeigths = rowMeans(sapply(tfLayerPropertiesSet, function(tfld) tfld[, 'dynamic_norm']))/
                  rowMeans(sapply(tfLayerPropertiesSet, function(tfld) tfld[, 'nWin_total_norm']))
hClusterReordered = reorder(hCluster, wts = reorderWeigths, agglo.FUN = 'mean')

```

## Heatmap

Creates the heatmap shown in panel 5E.

```{r}

colnames(tfLayerNRegScaled) = c('I', 'II', 'III', 'IV', 'V', 'VI', 'VII')
colnames(tfLayerDynamicsScaled) = c('I', 'II', 'III', 'IV', 'V', 'VI', 'VII')

customColorRamp = function(layerData){
  return(colorRamp2(c(min(layerData), (min(layerData) + max(layerData))/2, max(layerData)), c('#0000FF', '#000000', '#FFFF00')))
}

customColorRampCyanRed = function(layerData){
  return(colorRamp2(c(min(layerData), (min(layerData) + max(layerData))/2, max(layerData)), c('#00FFFF', '#000000', '#FF0000')))
}

Heatmap(tfLayerDynamicsScaled, name ='Dynamicity index (Z-score)', cluster_columns = FALSE, cluster_rows = hClusterReordered, show_row_names = FALSE,
        col = customColorRamp(tfLayerDynamicsScaled),
        heatmap_legend_param = list(color_bar = "continuous")) +
  Heatmap(tfLayerNRegScaled, name ='Number of bound regions (Z-score)', cluster_columns = FALSE, cluster_rows = hClusterReordered,
          col = customColorRampCyanRed(tfLayerNRegScaled),
          heatmap_legend_param = list(color_bar = "continuous")) +
  Heatmap(tfLayerClassif, name = 'Classification', cluster_rows = hClusterReordered, show_column_names = FALSE, col = c("#a2cd5a", "#4682b4", "#ff3030")) 

```

## Functional set enrichment analysis

Given our ranked TF list, we assess the significance for pioneers at the top and migrants in the bottom.

```{r, message = FALSE}

library('fgsea')
library('ggplot2')
library(gridExtra)

# Extract TF names based on their classification
tfSortedByClassif = list('p' = tfLayerPropertiesSet[[1]]$realName[tfLayerClassif == 'p'],
                         's' = tfLayerPropertiesSet[[1]]$realName[tfLayerClassif == 's'],
                         'm' = tfLayerPropertiesSet[[1]]$realName[tfLayerClassif == 'm'])

# Ranked list for FSEA
tfRankedByDynNReg = rev(seq(length(tfLayerClassif)))
names(tfRankedByDynNReg) = tfLayerPropertiesSet[[1]]$realName[hClusterReordered$order]
fseaResult = fgsea(pathways = tfSortedByClassif, stats = tfRankedByDynNReg, nperm = 1e5, gseaParam = 0)

# Plotting!
ePioneer = plotEnrichment(tfSortedByClassif[['p']], tfRankedByDynNReg, gseaParam = 0) + labs(title = 'Pioneers')
eSettler = plotEnrichment(tfSortedByClassif[['s']], tfRankedByDynNReg, gseaParam = 0) + labs(title = 'Settlers')
eMigrant = plotEnrichment(tfSortedByClassif[['m']], tfRankedByDynNReg, gseaParam = 0) + labs(title = 'Migrants')

grid.arrange(ePioneer, eSettler, eMigrant)

```

# TF interactions overlap

In order to investigate commonalities and specifities for TF interactions happening at genes in different gene modules, we plotted Venn diagrams corresponding to the edges observed in each one of the networks corresponding to a different gene module using the *Vennerable* package. This packages allows for the construction of diagrams which regions areas are proportional to the number of elements in the set they represent, called Euler diagrams [@chow2003]. 

## Assemblying edge sets and visualization

We implicitly assign an ID to each edge depending of its position in the adjacency matrix using the 'which' function. With this method, corresponding edges in different networks will have the same ID (assuming the adjacency matrix for all networks correspond to the same set of nodes and are properly ordered).

The sets are plotted using the *Vennerable* package. We import a file with two auxiliary functions. The first one, *plotEulerDiagram*, creates the diagram from the a list of edge IDs with custom plotting style. The second one is an override from an internal Vennerable function that fix it for the case of 7-set Euler diagrams.

```{r, warning = FALSE}

library(Vennerable)
source(paste0(srcFolder, '/5F_VENNERABLE.R'))

# Selecting useful edges
edgeIds = lapply(seq(length(tfGraphFilteredAdjMatList)),
                 function(c) which(as.vector(tfGraphFilteredAdjMatList[[c]]) > 0))

plotEulerDiagram(edgeIds)
```

## Layered analysis of TF interaction overlap

Since we observed that TFs at the bottom of the hierarchy bind a specific set of regions, we decided to analyze how their interactions are shared among different transcriptional gene modules.

```{r, warning = FALSE}

# Top-outgoing edges
topEdgeIds = lapply(seq(length(tfGraphFilteredAdjMatList)), function(c){
  tfsInTop = tfLayerPropertiesSet[[c]]$layer == 'top'
  tfGraphFilteredAdjMatList[[c]][!tfsInTop,] = 0 # removing all edges that don't com out of a top layer TF
  return(which(as.vector(tfGraphFilteredAdjMatList[[c]]) > 0))
})

# Neither top-outgoing or bottom-incoming edges
coreEdgeIds = lapply(seq(length(tfGraphFilteredAdjMatList)), function(c){
  tfsInTop = tfLayerPropertiesSet[[c]]$layer == 'top'
  tfGraphFilteredAdjMatList[[c]][!tfsInTop,] = 0 # removing all edges that come out of a top layer TF
  tfsInBottom = tfLayerPropertiesSet[[c]]$layer == 'bottom'
  tfGraphFilteredAdjMatList[[c]][,tfsInBottom] = 0 # removing all edges that incide over a bottom layerd TF
  return(which(as.vector(tfGraphFilteredAdjMatList[[c]]) > 0))
})

# Bottom-incoming edges
bottomEdgeIds = lapply(seq(length(tfGraphFilteredAdjMatList)), function(c){
  tfsInBottom = tfLayerPropertiesSet[[c]]$layer == 'bottom'
  tfGraphFilteredAdjMatList[[c]][,!tfsInBottom] = 0 # removing all edges that don't incide over a bottom layerd TF
  return(which(as.vector(tfGraphFilteredAdjMatList[[c]]) > 0))
})

# Plotting!
edTop = plotEulerDiagram(topEdgeIds)
edCore = plotEulerDiagram(coreEdgeIds)
edBottom = plotEulerDiagram(bottomEdgeIds)
```

# Session Info
```{r}
sessionInfo()
```

# References